### DISCUSSION TITLE
Mobile use

### DISCUSSION THREAD

**Discussion by vagstar** [dscripka/openWakeWord#30] [Q&A]
Hey, is it possible to convert this to Core ML format using CoreML tools by Apple to use with iOS apps? 

Conversion from ONNX format seems to no longer be supported by CoreMLtools. Would it be possible to have these models in any other format supported by TensorFlow or PyTorch?

**Comment by dscripka** [comment]
Good point, @vagstar. I initially chose ONNX because it seemed to have enough support by other frameworks to enable exactly these types of conversions. With CoreMLtools removing that support, it does get more complicated.

However, I have had luck with the also deprecated, but still functional [onnx-tensorflow library](https://github.com/onnx/onnx-tensorflow). For example, this will convert the audio embedding model from onnx to a Tensorflow SavedModel directory:

```
import onnx
from onnx_tf.backend import prepare

onnx_model = onnx.load('embedding_model.onnx') 
tf_rep = prepare(onnx_model, device="CPU")

tf_rep.export_graph("embedding_model_tf_from_onnx")
```

In theory, you can then convert this with CoreMLtools as discussed [here](https://coremltools.readme.io/docs/tensorflow-2).

However, I do also have the original Tensorflow/Pytorch implementations of the models. If the above approach doesn't work for you, let me know I can try to include the base models as artifacts for future releases of openWakeWord.

**Reply by vagstar** [reply]
Thanks for taking the time to reply to my query. Can you please give a description of the input and output such as shape, type? Since the model takes frames of 80ms with 16kHz and single channel, I'm assuming input shape to be (1280, 1, 1). Is this correct?

**Reply by dscripka** [reply]
There are actually three different models in openWakeWord. You can check the input and output shapes of the models via the ONNX attributes after loading them in openWakeWord. For example:

```
# Melspectrogram model
oww.preprocessor.melspec_model.get_inputs()[0].shape
oww.preprocessor.melspec_model.get_outputs()[0].shape

# Embedding model
oww.preprocessor.embedding_model.get_inputs()[0].shape
oww.preprocessor.embedding_model.get_outputs()[0].shape

# Classification (wakeword) models
oww.models['hey_mycroft_v4'].get_inputs()[0].shape
oww.models['hey_mycroft_v4'].get_outputs()[0].shape
```

The inputs to the melspectrogram model are 16-bit integers values (raw audio values). For the embedding and classification models, the inputs are all 32-bit floats.
